### PR TITLE
PoolState, and price computation

### PR DESCRIPTION
**Pull request by dorimedini-starkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/892)

### CODE REVIEW

**Review by amosStarkware — CHANGES_REQUESTED** [review]
Please rebase, for some reason it took the reviewer really long to review this

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/892#-:-NeIAlrE9nOj7b62fXR4:bvzbt88)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @dorimedini-starkware)

___
*[`crates/blockifier/src/fee/gas_usage.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeHhCL9C4CHsar-Ji23:-NeHhCLA-4t8rhYhLMyI:b-s2fb8w) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage.rs#L26)):*
> ```XML
> 
> impl PoolState {
>     pub fn tvl_in_wei(&self) -> u128 {
> ```

shouldn't this be a getter? WYT?



_Code quote (i):_
```XML
pub fn tvl_in_wei(&self) -> u128 {
```



_Suggestion (ii):_
```XML
pub fn get_tvl_in_wei(&self) -> u128 {
```
___
*[`crates/blockifier/src/fee/gas_usage.rs` line 35 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI1zeb2LBXB3vfL5Y0:-NeI1zeb2LBXB3vfL5Y1:b4xztht) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage.rs#L35)):*
> ```XML
> /// Returns the weighted average Wei / STRK ratio, given the states of a collection of pools.
> /// Pools are weighted by TVL in units of Wei.
> pub fn get_wei_to_strk_ratio_from_pool_states(pool_states: &[PoolState]) -> f64 {
> ```

Is using f64 good enough?
it only holds numbers with up to 15-16 without losing information
https://stackoverflow.com/questions/65719216/why-does-rust-only-use-16-significant-digits-for-f64-equality-checks#:~:text=An%20f64%2C%20as,are%20the%20same.
it may be good enough but I thought I should bring it up

also, what do you think of renaming to `get_median_wei_to_strk_ratio`?



_Code quote:_
```XML
pub fn get_wei_to_strk_ratio_from_pool_states(pool_states: &[PoolState]) -> f64 {
```
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 14 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI3s9I2EtvysqrJ4oW:-NeI3s9I2EtvysqrJ4oX:bgqfuxx) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L14)):*
> ```XML
> macro_rules! assert_f64_eq {
>     ($left:expr, $right:expr) => {
>         match (&$left, &$right) {
> ```

why is the match expression needed?



_Code quote:_
```XML
atch (&$left, &$right) {
```
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 19 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI4due48mQo1zS9rpa:-NeI4duf8_5EMbImDrxs:b-tor44i) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L19)):*
> ```XML
>                     panic!(
>                         "assertion failed: `(left == right)`\n  left: `{:?}`,\n right: `{:?}`",
>                         &*left_val, &*right_val
> ```

can you explain?



_Code quote:_
```XML
&*left_val, &*right_val
```
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 138 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeJ5sZa-hQ24mt6xeNq:-NeJ5sZa-hQ24mt6xeNr:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L138)):*
> ```XML
> /// Sanity tests for STRK<->ETH price computation.
> #[test]
> fn test_pool_state() {
> ```





_Code quote (i):_
```XML
fn test_pool_state() {
```



_Suggestion (ii):_
```XML
fn test_get_wei_to_strk_ratio_from_pool_states() {
```

**Review by dorimedini-starkware — COMMENTED** [review]
lol

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/892#-:-NeSyPf94C6LgU7PQH8t:b2ni78k)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @amosStarkware and @dorimedini-starkware)

___
*[`crates/blockifier/src/fee/gas_usage.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeHhCL9C4CHsar-Ji23:-NeSv9AFBIj5qM22khsu:b-biyxsc) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage.rs#L26)):*
<details><summary><i>Previously, amosStarkware wrote…</i></summary><blockquote>

shouldn't this be a getter? WYT?
</blockquote></details>

IMO no, because it is not direct access to a field - it is a (silly) computation
___
*[`crates/blockifier/src/fee/gas_usage.rs` line 35 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI1zeb2LBXB3vfL5Y0:-NeSvFW-33skrjlUJH-w:b-warruw) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage.rs#L35)):*
<details><summary><i>Previously, amosStarkware wrote…</i></summary><blockquote>

Is using f64 good enough?
it only holds numbers with up to 15-16 without losing information
https://stackoverflow.com/questions/65719216/why-does-rust-only-use-16-significant-digits-for-f64-equality-checks#:~:text=An%20f64%2C%20as,are%20the%20same.
it may be good enough but I thought I should bring it up

also, what do you think of renaming to `get_median_wei_to_strk_ratio`?
</blockquote></details>

1. I agree that it is not exact; but the 15-16 precision digits is because [53 bits are used for precision](https://en.wikipedia.org/wiki/Double-precision_floating-point_format#IEEE_754_double-precision_binary_floating-point_format:_binary64). As a ratio value used for approximate conversion, this may be good enough; or, we can decide to only implement `wei_to_strk(pool_states: &[PoolState], wei: u128) -> u128` function that only converts actual prices (changed to that now). WDYT?
2. Rename will be needed when that's what the function does
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 14 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI3s9I2EtvysqrJ4oW:-NeSwqWcD_VFRYYiXct0:b-x99525) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L14)):*
<details><summary><i>Previously, amosStarkware wrote…</i></summary><blockquote>

why is the match expression needed?
</blockquote></details>

both `left_val` and `right_val` are used twice; so when we write
```
assert_f64_eq!(get_wei_to_strk_ratio_from_pool_states(&[state_0, state_1]), expected_weighted);
```
then `get_wei_to_strk_ratio_from_pool_states(&[state_0, state_1])` is written twice. More specifically, `&[state_0, state_1]` is created twice, which rust doesn't like (`state_0` is moved, `state_1` is also moved).
The `match` solves this by borrowing.
See [the implementation](https://doc.rust-lang.org/src/core/macros/mod.rs.html#36) of `assert_eq!` (I copied from there)
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 19 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeI4due48mQo1zS9rpa:-NeSxoz65eADxq6eJnH6:bf4xper) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L19)):*
<details><summary><i>Previously, amosStarkware wrote…</i></summary><blockquote>

can you explain?
</blockquote></details>

[I had to google this, actually](https://stackoverflow.com/questions/41273041/what-does-combined-together-do-in-rust#answer-41273406)
tl;dr:
"the `*` triggers an explicit deref, which can be overloaded via [`ops::Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html)"
Not actually needed in this case (no overloaded deref), but makes macro more future-proof
___
*[`crates/blockifier/src/fee/gas_usage_test.rs` line 138 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/892#-NeJ5sZa-hQ24mt6xeNq:-NeSyOCtEhy9pP6cLKQp:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/bc37c1709431d7560f3a8e3ef4fb51d21951af8c/crates/blockifier/src/fee/gas_usage_test.rs#L138)):*
> ```XML
> /// Sanity tests for STRK<->ETH price computation.
> #[test]
> fn test_pool_state() {
> ```

Done.
